sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -116,6 +116,57 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     }
   }
 
+  override def visitCompoundOrSingleStatement(
+      ctx: CompoundOrSingleStatementContext): CompoundBody = {
+    if (ctx.singleCompound() != null) {
+      visit(ctx.singleCompound()).asInstanceOf[CompoundBody]
+    } else {
+      val logicalPlan = visitSingleStatement(ctx.singleStatement())
+      CompoundBody(List(SparkStatementWithPlan(
+        parsedPlan = logicalPlan,
+        sourceStart = ctx.start.getStartIndex,
+        sourceEnd = ctx.stop.getStopIndex + 1)))
+    }
+  }
+
+  override def visitSingleCompound(ctx: SingleCompoundContext): CompoundBody = {
+    visit(ctx.beginEndCompoundBlock()).asInstanceOf[CompoundBody]
+  }
+
+  private def visitCompoundBodyImpl(ctx: CompoundBodyContext): CompoundBody = {
+    val buff = ListBuffer[CompoundPlanStatement]()
+    for (i <- 0 until ctx.getChildCount) {
+      val child = ctx.getChild(i)
+      val visitedChild = visit(ctx.getChild(i))
+      visitedChild match {
+        case statement: CompoundPlanStatement => buff += statement
+        case null if child.isInstanceOf[TerminalNodeImpl] =>
```

Comment:
why would the child not be an instance of `TerminalNodeImpl`? If this is enforced by the parser, then you can just remove the check

Reply:
Well, I agree. We had only `case null =>` but it was very unclear what it meant, so we decided to add this check. 

This way, if anything changes in the future, we would get exceptions in our tests if this `null` does not happen because the `child` is instance of `TerminalNodeImpl`.

Not sure if this is the best way to do it though. I guess that it's probably better to not visit `child` at all if it's a terminal node. I don't know how better to achieve this. Or maybe we are fine with leaving only `case null =>` and adding comment that explains it?

Reply:
Do you have an example SQL to trigger this null case?

Reply:
This basically happens within every SQL statement - `TerminalNodeImpl` corresponds to keywords from the SQL grammar, so for example if the query is like:
<img width="328" alt="image" src="https://github.com/apache/spark/assets/163021185/3f033dd4-0fea-4274-883e-a40f910eb9d9">

Then, on the top level, `TerminalNodeImpl` are semicolons:
<img width="328" alt="image" src="https://github.com/apache/spark/assets/163021185/f293b22c-609b-47e0-945e-69507970286f">

And then deeper in the tree it could be a `SELECT`:
<img width="1060" alt="image" src="https://github.com/apache/spark/assets/163021185/d4636fa3-c6db-4771-950f-a915bca8aa74">

Or whichever other keyword.

Reply:
Should be resolved with the latest changes in `visitCompoundBodyImpl`.